{
    "id": "SOLR-5402",
    "title": "SolrCloud 4.5 bulk add errors in cloud setup",
    "details": {
        "affect_versions": "4.5,                                            4.5.1",
        "status": "Closed",
        "fix_versions": [
            "4.6",
            "4.7",
            "6.0"
        ],
        "components": [
            "SolrCloud"
        ],
        "type": "Bug",
        "priority": "Major",
        "labels": "",
        "resolution": "Fixed"
    },
    "description": "We use out of the box Solr 4.5.1 no customization done. If we merge documents via SolrJ to a single server it is perfectly working fine.\n\nBut as soon as we add another node to the cloud we are getting following while merging documents. We merge about 500 at a time using SolrJ. These 500 documents in total are about few MB (1-3) in size.\n\nThis is the error we are getting on the server (10.10.10.116 - IP is irrelavent just for clarity)where merging is happening. 10.10.10.119 is the new node here. This server gets RemoteSolrException\n\nshard update error StdNode: http://10.10.10.119:8980/solr/mycore/:org.apache.solr.client.solrj.impl.HttpSolrServer$RemoteSolrException: Illegal to have multiple roots (start tag in epilog?).\n at [row,col \n{unknown-source}]: [1,12468]\n\tat org.apache.solr.client.solrj.impl.HttpSolrServer.request(HttpSolrServer.java:425)\n\tat org.apache.solr.client.solrj.impl.HttpSolrServer.request(HttpSolrServer.java:180)\n\tat org.apache.solr.update.SolrCmdDistributor$1.call(SolrCmdDistributor.java:401)\n\tat org.apache.solr.update.SolrCmdDistributor$1.call(SolrCmdDistributor.java:1)\n\tat java.util.concurrent.FutureTask$Sync.innerRun(Unknown Source)\n\tat java.util.concurrent.FutureTask.run(Unknown Source)\n\tat java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)\n\tat java.util.concurrent.FutureTask$Sync.innerRun(Unknown Source)\n\tat java.util.concurrent.FutureTask.run(Unknown Source)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.runTask(Unknown Source)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)\n\tat java.lang.Thread.run(Unknown Source)\n\n\nOn the other server 10.10.10.119 we get following error\n\norg.apache.solr.common.SolrException: Illegal to have multiple roots (start tag in epilog?).\n at [row,col {unknown-source}\n]: [1,12468]\n\tat org.apache.solr.handler.loader.XMLLoader.load(XMLLoader.java:176)\n\tat org.apache.solr.handler.UpdateRequestHandler$1.load(UpdateRequestHandler.java:92)\n\tat org.apache.solr.handler.ContentStreamHandlerBase.handleRequestBody(ContentStreamHandlerBase.java:74)\n\tat org.apache.solr.handler.RequestHandlerBase.handleRequest(RequestHandlerBase.java:135)\n\tat org.apache.solr.core.SolrCore.execute(SolrCore.java:1859)\n\tat org.apache.solr.servlet.SolrDispatchFilter.execute(SolrDispatchFilter.java:703)\n\tat org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:406)\n\tat org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:195)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:243)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:210)\n\tat org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:222)\n\tat org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:123)\n\tat org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:171)\n\tat org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:99)\n\tat org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:936)\n\tat org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:118)\n\tat org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:407)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1004)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:589)\n\tat org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:310)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)\n\tat java.lang.Thread.run(Thread.java:662)\nCaused by: com.ctc.wstx.exc.WstxParsingException: Illegal to have multiple roots (start tag in epilog?).\n at [row,col \n{unknown-source}\n]: [1,12369]\n\tat com.ctc.wstx.sr.StreamScanner.constructWfcException(StreamScanner.java:630)\n\tat com.ctc.wstx.sr.StreamScanner.throwParseError(StreamScanner.java:461)\n\tat com.ctc.wstx.sr.BasicStreamReader.handleExtraRoot(BasicStreamReader.java:2155)\n\tat com.ctc.wstx.sr.BasicStreamReader.nextFromProlog(BasicStreamReader.java:2070)\n\tat com.ctc.wstx.sr.BasicStreamReader.nextFromTree(BasicStreamReader.java:2647)\n\tat com.ctc.wstx.sr.BasicStreamReader.next(BasicStreamReader.java:1019)\n\tat org.apache.solr.handler.loader.XMLLoader.processUpdate(XMLLoader.java:213)\n\tat org.apache.solr.handler.loader.XMLLoader.load(XMLLoader.java:174)\n\t... 22 more\n\n\nWe tried with 4.5.0 first and then with 4.5.1 later. Both produce exact same error. \n\nNote: This error is not present when we switched exact setup to 4.4.0",
    "attachments": {},
    "issue_links": {},
    "comments": [
        {
            "author": "Sai Gadde",
            "id": "comment-13808666",
            "date": "2013-10-30T01:53:02+0000",
            "content": "http://www.mail-archive.com/solr-user@lucene.apache.org/msg90748.html\nwe already tried tomcat setting from this post size to a very high value of 100MB. We used exact setting sepecified in the above thread.\n\nhttps://issues.apache.org/jira/browse/SOLR-5331\nformdataUploadLimitInKB and multipartUploadLimitInKB limits are set to very high value 2GB. As suggested in the following thread.\n\nThese changes did not help resolving this issue.  SOLR-5331 is closed so, opening a new one. "
        },
        {
            "author": "Michael Tracey",
            "id": "comment-13809099",
            "date": "2013-10-30T13:58:55+0000",
            "content": "I can confirm this issue on Jetty, both the bundled 8.x version from the example directory, and a 9.0.6 upgraded version of jetty with Solr 4.5.1.  I believe the issue is separate from the container. "
        },
        {
            "author": "Mark Miller",
            "id": "comment-13809341",
            "date": "2013-10-30T17:03:18+0000",
            "content": "Have you tried fewer than 500 docs at a time to see if the problem persists? "
        },
        {
            "author": "Michael Tracey",
            "id": "comment-13809355",
            "date": "2013-10-30T17:19:00+0000",
            "content": "Yes, I've found that a 50 docs will trigger the problem. "
        },
        {
            "author": "Sai Gadde",
            "id": "comment-13809833",
            "date": "2013-10-31T01:30:09+0000",
            "content": "I even tried adding 1 document at a time. Still the problem is present.\n\nIf there is only one server there no errors, everything works fine. updating even 500 documents is fine. \n\nBut when another node comes online SolrCmdDistributor gets these exceptions from remote server response. New Solr node also prints it's own error stack as explained in the report above.\n\nThis exact same setup works without issues in 4.4.0. I tried with both 4.5.0 and 4.5.1 and both produce same errors "
        },
        {
            "author": "Michael Tracey",
            "id": "comment-13810551",
            "date": "2013-10-31T18:55:50+0000",
            "content": "I can confirm the same thing, it errors with a single document.  That's the last test I did before rolling my infrastructure back to 4.4.  I can index 1000's of documents at a time without issue to a single server, but when SolrCloud 4.5.1 tries to sync two nodes, it fails (same errors as above). "
        },
        {
            "author": "Greg Walters",
            "id": "comment-13822728",
            "date": "2013-11-14T18:30:17+0000",
            "content": "I've been able to reproduce this issue using curl to add documents but using the post.jar provided in the Solr example or using a solrj client I'm unable to reproduce this issue having tried batches up to 5000 documents. "
        },
        {
            "author": "Dave Seltzer",
            "id": "comment-13824158",
            "date": "2013-11-15T22:07:36+0000",
            "content": "We're having the same issue over here. I was able to insert hundreds of documents per \"add\" using the REST API and a single collection.\n\nWhen I re-created my collection as as a cloud instance with NumShards=2, It fails whenever I try to insert more than one document per \"add\" request. "
        },
        {
            "author": "Mark Miller",
            "id": "comment-13826110",
            "date": "2013-11-19T03:07:56+0000",
            "content": "Sorry I have not had time to help on this one. It's on my list to look into. "
        },
        {
            "author": "Mark Miller",
            "id": "comment-13827109",
            "date": "2013-11-19T23:44:26+0000",
            "content": "FYI, 4.6 no longer sends updates via XML - so I have not looked into this at all yet, but there is some chance the problem has been sidestepped. "
        },
        {
            "author": "Mark Miller",
            "id": "comment-13881381",
            "date": "2014-01-24T20:18:03+0000",
            "content": "pretty sure this issue no longer exists. If it's spotted again we can reopen. "
        }
    ]
}